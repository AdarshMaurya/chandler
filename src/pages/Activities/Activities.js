import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import {View, FlatList, ActivityIndicator, Text} from 'react-native';

import {
  Navbar,
  EmptyActivity,
  LastTwoYearsStatistics,
  YearChart,
} from 'components';
import {commonStyles} from 'theme';
import {styles} from './Activities.styles';

export class Activities extends PureComponent {
  static propTypes = {
    back: PropTypes.func.isRequired,
    getActivitiesByContact: PropTypes.func.isRequired,
    contact: PropTypes.object.isRequired,
    activities: PropTypes.array.isRequired,
    isFetching: PropTypes.bool.isRequired,
  };

  componentWillMount() {
    this.props.getActivitiesByContact();
  }

  keyExtractor = (item, index) => String(item.id);

  renderHeader = () => {
    const {isFetching} = this.props;

    if (isFetching) {
      return null;
    }

    return (
      <View style={styles.headerContainer}>
        <LastTwoYearsStatistics
          image={require('assets/icons/activities.png')}
          title1="in 2017"
          count1={3}
          title2="in 2016"
          count2={6}
        />

        <YearChart />
      </View>
    );
  };

  renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  renderItem = ({item, index}) => {
    const {activities} = this.props;
    const activity = activities[index];

    return (
      <View style={styles.activityContainer}>
        <View style={commonStyles.row}>
          <Text style={styles.textLeft}>Activity</Text>
          <View style={commonStyles.flex} />
          <Text style={styles.textRight}>
            {moment(activity.date_it_happened).fromNow()}
          </Text>
        </View>
        {activity.description && (
          <Text style={styles.textInfo}>{activity.description}</Text>
        )}
      </View>
    );
  };

  render() {
    const {
      back,
      activities,
      getActivitiesByContact,
      isFetching,
      contact,
    } = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title="Activities" onBack={back} />
        {isFetching || activities.length ? (
          <FlatList
            data={activities}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListHeaderComponent={this.renderHeader}
            ListFooterComponent={this.renderFooter}
            onEndReached={getActivitiesByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-activities.png')}
            title={
              'Have you done an activity with ' +
              contact.first_name +
              ' lately?'
            }
            subtitle="Keep track of what youâ€™ve done."
          />
        )}
      </View>
    );
  }
}
