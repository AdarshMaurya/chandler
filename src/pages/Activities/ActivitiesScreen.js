import {connect} from 'react-redux';

import {Activities} from './Activities';
import {back} from 'redux/router';

export const mapStateToProps = (state, {navigation}) => ({});

export const mapDispatchToProps = dispatch => ({
  back: () => dispatch(back()),
});

export const ActivitiesScreen = connect(mapStateToProps, mapDispatchToProps)(
  Activities,
);
