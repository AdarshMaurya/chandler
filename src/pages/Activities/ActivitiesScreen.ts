import {IRootState} from '@models';
import {connect} from 'react-redux';

import {getActivitiesByContact} from '@redux/activities';
import {back} from '@redux/router';
import {Activities} from './Activities';

export const mapStateToProps = (state: IRootState, {navigation}) => {
  const contact = state.contacts[navigation.state.params];
  return {
    contact,
    activities: (contact.activities || []).map(
      activityId => state.activities[activityId]
    ),
    isFetching: state.getActivitiesByContact.isFetching,
    statistics: state.getActivitiesByContact.statistics,
  };
};

export const mapDispatchToProps = (dispatch, {navigation}) => ({
  back: () => dispatch(back()),
  getActivitiesByContact: () =>
    dispatch(getActivitiesByContact(navigation.state.params)),
});

export const ActivitiesScreen = connect(mapStateToProps, mapDispatchToProps)(
  Activities
);
