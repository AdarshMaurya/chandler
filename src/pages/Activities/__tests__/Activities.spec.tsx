import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {Activities} from '../Activities';

describe('Pages', () => {
  describe('Activities', () => {
    let back;
    let getActivitiesByContact;
    const contact = {first_name: 'Theo'};

    beforeEach(() => {
      back = jest.fn();
      getActivitiesByContact = jest.fn();
    });

    it('should renders correctly without activities and fetching', () => {
      const activities = [];
      const tree = shallow(
        <Activities
          contact={contact as any}
          back={back}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly without activities and not fetching', () => {
      const activities = [];
      const tree = shallow(
        <Activities
          back={back}
          contact={contact as any}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly with activities and fetching', () => {
      const activities = [
        {description: 'My description', date_it_happened: '1991-12-17'},
      ];
      const tree = shallow(
        <Activities
          contact={contact as any}
          back={back}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
    });

    it('should renders correctly with activities and fetching', () => {
      const activities = [
        {description: 'My description', date_it_happened: '1991-12-17'},
        {description: 'My description 2', date_it_happened: '1993-12-17'},
      ];
      const tree = shallow(
        <Activities
          contact={contact as any}
          back={back}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 1})))
      ).toMatchSnapshot();
    });

    it('should display the activity indicator in the footer if it is fetching', () => {
      const activities = [
        {description: 'My description', date_it_happened: '1991-12-17'},
        {description: 'My description 2', date_it_happened: '1993-12-17'},
      ];
      const tree = shallow(
        <Activities
          contact={contact as any}
          back={back}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities as any}
        />
      );
      expect(toJson(shallow(tree.instance().renderFooter()))).toMatchSnapshot();
    });

    it('should not display the activity indicator in the footer if it is not fetching', () => {
      const activities = [
        {description: 'My description', date_it_happened: '1991-12-17'},
        {description: 'My description 2', date_it_happened: '1993-12-17'},
      ];
      const tree = shallow(
        <Activities
          contact={contact as any}
          back={back}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
        />
      );
      expect(tree.instance().renderFooter()).toBeNull();
    });

    it('should have a keyExtractor', () => {
      const activities = [{id: 1}, {id: 2}];
      const tree = shallow(
        <Activities
          contact={contact as any}
          back={back}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
        />
      );
      expect(tree.instance().keyExtractor(activities[0])).toBe('1');
      expect(tree.instance().keyExtractor(activities[1])).toBe('2');
    });
  });
});
