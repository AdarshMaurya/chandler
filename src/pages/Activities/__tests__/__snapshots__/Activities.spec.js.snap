// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages Activities should display the activity indicator in the footer if it is fetching 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "marginVertical": 20,
    }
  }
/>
`;

exports[`Pages Activities should renders correctly with activities and fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <FlatList
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={
      Array [
        Object {
          "date_it_happened": "1991-12-17",
          "description": "My description",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Pages Activities should renders correctly with activities and fetching 2`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
        }
      }
    >
      Activity
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#97A5B2",
        }
      }
    >
      25 years ago
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 14,
        "marginTop": 5,
      }
    }
  >
    My description
  </Text>
</View>
`;

exports[`Pages Activities should renders correctly with activities and fetching 3`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <FlatList
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={
      Array [
        Object {
          "date_it_happened": "1991-12-17",
          "description": "My description",
        },
        Object {
          "date_it_happened": "1993-12-17",
          "description": "My description 2",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Pages Activities should renders correctly with activities and fetching 4`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
        }
      }
    >
      Activity
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#97A5B2",
        }
      }
    >
      25 years ago
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 14,
        "marginTop": 5,
      }
    }
  >
    My description
  </Text>
</View>
`;

exports[`Pages Activities should renders correctly with activities and fetching 5`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
        }
      }
    >
      Activity
    </Text>
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#97A5B2",
        }
      }
    >
      23 years ago
    </Text>
  </View>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 14,
        "marginTop": 5,
      }
    }
  >
    My description 2
  </Text>
</View>
`;

exports[`Pages Activities should renders correctly without activities and fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <FlatList
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</View>
`;

exports[`Pages Activities should renders correctly without activities and not fetching 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <EmptyActivity
    image={1}
    subtitle="Keep track of what youâ€™ve done."
    title="Have you done an activity with Theo lately?"
  />
</View>
`;
